bundles/org.openhab.core.config.discovery.usbserial.otheros/src/main/java/org/openhab/core/config/discovery/usbserial/otheros/internal/OtherOSUsbSerialDiscovery.java

```diff
@@ -0,0 +1,221 @@
+        for (UsbSerialDeviceInformation deviceInfo : lastScanResult) {
+            listener.usbSerialDeviceDiscovered(deviceInfo);
+        }
+    }
+
+    @Override
+    public void unregisterDiscoveryListener(UsbSerialDiscoveryListener listener) {
+        discoveryListeners.remove(listener);
+    }
+
+    /**
+     * Traverse the USB tree for devices that are children of the ROOT hub, and return a set of USB device information.
+     * 
+     * @param usbHub the hub whose children are to be found.
+     * @return a set of USB device information.
+     */
+    private Set<UsbSerialDeviceInformation> scanAllUsbDevicesInformation() {
+        try {
+            return scanChildUsbDeviceInformation(UsbHostManager.getUsbServices().getRootUsbHub());
+        } catch (SecurityException | UsbException e) {
+            logger.warn("Error getting USB device information: {}", e.getMessage());
+            return Set.of();
+        }
+    }
+
+    /**
+     * Traverse the USB tree for devices that are children of the given hub, and return a set of USB device information.
+     * 
+     * @param usbHub the hub whose children are to be found.
+     * @return a set of USB device information.
+     */
+    private Set<UsbSerialDeviceInformation> scanChildUsbDeviceInformation(UsbHub usbHub) {
+        Set<UsbSerialDeviceInformation> result = new HashSet<>();
+
+        @SuppressWarnings("unchecked")
+        List<UsbDevice> deviceList = usbHub.getAttachedUsbDevices();
+
+        deviceList.forEach(usbDevice -> {
+            if (usbDevice.isUsbHub()) {
+                result.addAll(scanChildUsbDeviceInformation((UsbHub) usbDevice));
+            } else {
+                UsbDeviceDescriptor d = usbDevice.getUsbDeviceDescriptor();
+                short vendorId = d.idVendor();
+                short productId = d.idProduct();
+
+                String manufacturer = null;
+                String product = null;
+                String serialNumber = null;
+
+                /*
+                 * Note: the getString() calls below may fail depending on the Operating System:
+                 * - on Windows if no libusb device driver is installed for the device.
+                 * - on Linux if the user has no write permission on the USB device file.
+                 */
+                try {
+                    product = usbDevice.getString(d.iProduct());
+                    manufacturer = usbDevice.getString(d.iManufacturer());
+                    serialNumber = usbDevice.getString(d.iSerialNumber());
+                } catch (UnsupportedEncodingException | UsbDisconnectedException | UsbException e) {
+                    // ignore because this would be a 'normal' runtime failure
+                }
+
+                if (product == null || product.isBlank()) {
+                    product = UsbProductDatabase.getProduct(vendorId, productId);
+                }
+
+                /*
+                 * TODO are the following three data required (??)
```

Comment:
It is probably needed if used beyond the finder. My understanding is the serial port is at the interface level, not the device level. And there can be multiple interfaces in a configuration in a device.

Reply:
Yeah. But I think device level is sufficient for making addon suggestions.

Reply:
Agree, but if we make it available as a serial discovery, it needs all to be able to be used outside of the finder.

Reply:
This of course leads to the next question. The serial discovery classes focus on finding USB devices with serial interfaces. The finder, in theory could look broader than that (although I don't know if there is any current use). E.g. what about a camera or speaker directly connected through USB to the server. Most likely, discovery would be better done by detecting the process in that case, but in theory it could also be done by detecting the configurations and interfaces on the USB device. But then it wouldn't fit in this class anymore.

Reply:
Indeed.